File: src/sinks/dev.ts
import type { WideEvent } from '../core/schema.ts';
import type { Sink } from '../core/types.ts';

export type MemorySink = Sink & {
	clear: () => void;
	readonly events: WideEvent[];
};

const OUTCOME_MARKS: Record<string, string> = {
	cancelled: '⊘',
	error: '✗',
	success: '✓',
};

export function createConsolePrettySink(name = 'console'): Sink {
	return {
		name,
		send: (event) => {
			try {
				console.info(formatHeader(event), event);
			} catch {
				console.info('[observe]', event);
			}
		},
	};
}

export function createMemorySink(name = 'memory'): MemorySink {
	const events: WideEvent[] = [];

	return {
		clear: () => {
			events.length = 0;
		},
		events,
		name,
		send: (event) => {
			events.push(event);
		},
	};
}

function formatHeader(event: WideEvent): string {
	const mark = OUTCOME_MARKS[event.outcome] ?? '?';
	const trace = event.trace_id.slice(0, 8);
	const span = event.span_id.slice(0, 8);
	return `[observe] ${mark} ${event.event} ${event.duration_ms}ms trace=${trace} span=${span}`;
}

File: python/observe_py/sinks/dev.py
"""Local dev + test sinks — twin of ``src/sinks/dev.ts``."""

from __future__ import annotations

from typing import Any

_OUTCOME_MARKS = {"cancelled": "⊘", "error": "✗", "success": "✓"}


class ConsolePrettySink:
    def __init__(self, name: str = "console") -> None:
        self.name = name

    def send(self, event: dict[str, Any]) -> None:
        try:
            print(_format_header(event), event)
        except Exception:
            print("[observe]", event)


class MemorySink:
    def __init__(self, name: str = "memory") -> None:
        self.name = name
        self.events: list[dict[str, Any]] = []

    def clear(self) -> None:
        self.events.clear()

    def send(self, event: dict[str, Any]) -> None:
        self.events.append(event)


def _format_header(event: dict[str, Any]) -> str:
    mark = _OUTCOME_MARKS.get(str(event.get("outcome")), "?")
    trace = str(event.get("trace_id", ""))[:8]
    span = str(event.get("span_id", ""))[:8]
    return (
        f"[observe] {mark} {event.get('event')} {event.get('duration_ms')}ms "
        f"trace={trace} span={span}"
    )
